### DISCUSSION TITLE
Request for Info: Wrapping Duckdb behind a Flask/API server for SOC2 compliance, which is accessible to data query tools(e.g. Dbeaver)

### DISCUSSION THREAD

**Discussion by RAbraham** [duckdb/duckdb#5376] [General]
Hi,
We have a DuckDB data file(size 2G and counting) which has sensitive personal identifying data. We are working towards SOC2 compliance and we also need to be HIPAA compliant.

Currently, we download the duckdb data file from an on-premise shared server(inside a VPN) onto our laptops and do our analysis. This is against SOC2 compliance. 

I'm wondering how has the community solved this requirement using DuckDB?

We can access the file via VPN via shared drive but I fear that this is sub-optimal? Every time a query is run on some editor(e.g. DBeaver or our automated evaluation script), I think the entire 2G file is downloaded perhaps to our laptop for every query which is not feasible.

One proposal being floated is to move away from DuckDb to PostgreSQL (Amazon RDS) but I wanted to try everything possible to stay on DuckDB if I can.

I was wondering if I could 
- spin up some flask server on the shared server (inside the VPN) which has direct access to the duckdb database. This server has an API which takes an SQL query, runs it on the duckdb data file and returns some data(e.g. json). This could work for some cases
- however our data scientists use tools like DBeaver etc. to run queries. I understand these editors require jdbc/odbc protocol set up? How can I make these editors/query tools talk to my flask server API? What would I have to do within my flask api to enable this? If so, are there any libraries that you know that can make this easy for me? or maybe this is a non-problem? If someone has done something like this before, I'd be grateful if you could point any resources to me.

If there are other solutions, please let me know.

**Comment by Dimpstone** [comment]
Rajiv,
DuckDB = In-Process (Embedded) Analytical Database.
You need Application/Service around the DuckDB:
[Remote Users] -- (Secure Data Transfer) -- [Local Data Processing : You Application/Service over DuckDB Engine]

**Reply by RAbraham** [reply]
Thanks @Dimpstone. Yeah, I'm aware of that pattern. I think @ajayaa 's answer is what I was looking for.

**Comment by ajayaa** [comment]
@RAbraham 

I don't think the entire 2GB file would be downloaded on each run of a query unless you are inserting data to the DB. Would be curious to know why you think the entire file is being downloaded on every query.

The other solution I can think of is to install Apache superset on the machine where duckdb file is present. You could log in to superset using oauth providers like github/google auth etc. However that would require your data scientists to use superset instead of DBeaver.

Reg. building a HTTP server on top of DuckDB, it won't work with DBeaver and all unless you write another layer of JDBC to HTTP. Instead of HTTP protocol what you are looking for is, a postgress client server protocol on top of duckdb. That would be too much of effort IMO.

**Reply by RAbraham** [reply]
@ajayaa you are right.
- though the first query on the db file on a remote server is around 5-10 seconds long, subsequent queries are very fast. similar queries on a local db file are sub second but this 5-10 secs delay could be very acceptable currently. 
- Apache Superset is a great idea and alternative as well. I'll check it out. 

Thank you very much for your advice!
